<p style="text-align: center;"><font size="8"><b>Section 1.6: Functions and Modules</b></font><br>


# Calling Functions

Pure functions exist as methods that can be called outside the context of a particular class. For example we've already seen the `round` and `len` functions. Remember that we call `round(a)`, not `a.round()`. Python provides many built-in functions. 

![functions](https://github.com/lukasbystricky/ISC-3313/blob/master/lectures/chapter2/images/functions.png?raw=true)

In [ ]:
pow(2,3)

8

In [ ]:
min(-2,-2.4,9.0)

-2.4

In [ ]:
ord("a")

97

# Print function

After performing a computation we need to see the result. In an interactive or notebook session we can simply type the name of the variable to see what it is.

In [ ]:
a = 1
a

1

When executing scripts however, we can no longer do this. Instead we use the `print` command (in fact we've already used this several times).

The `print` function is used to print information to the console. 

In [ ]:
print(a)
print("a")

1
a


Note the original Python 2 syntax, `print a`, (no parentheses) does not work in Python 3. However the syntax above (with parentheses) works in both Python 2 and Python 3.


Of course, `print` can be used to display more useful information. For example, suppose we have a variable called `t` that represents some length of time in seconds. We can use `print` to display not only `t`, but also the units by calling `print` with multiple arguments.

In [ ]:
t = 5.6
print(t, "s")

5.6 s


The `print` function automatically inserts a space between the two arguments. If we wish the avoid this, we can combine the two arguments into 1.

In [ ]:
print(str(t)+"s")

5.6s


Note that we have to convert `round(t,1)` to a string first before "adding" (the correct term is *concatenating*) it to `s`. The command

In [ ]:
print(t+"s")

TypeError: ignored

is illegal because we are attempting to add a `float` to a `str`.

Another option is to use the f-string formatting we talked about earlier,

In [ ]:
print(f"{t}s")

5.6s


# Modules

All the functions listed above are built-in to Python. This means they are automatically available once we start Python. But you'll notice that there's not actually all that many built-in functions. What if we want to take the logarithm of a number for example?

There are hundreds of other useful functions and classes that have been developed for Python which are not automatically loaded, but instead placed into specialized libraries called *modules* that can be individually loaded as needed.

If you installed Python using Anaconda, you will already have many modules installed. We will first take a look at the `math` module.

## Math module

The math module provides functions to do mathematical operations beyond addition, multiplication, exponentiation etc. For example suppose we want to take the cosine of a number. This is not built-in to Python, however a function called `cos` in the math library does this. 

To use the `cos` function we must import it from the math module. There are three possible ways to do this. Whichever method you use should be done at the beginning of your code (this is not strictly speaking necessary, but it is the most common placement, at the very least you have to do this *before* you use the function).

1) Import the entire module. 

In [ ]:
import math

At this point we still cannot use the `cos` command directly, we must specifically tell Python where the function is coming from using a *qualified name*.

In [ ]:
cos(2)

NameError: ignored

In [ ]:
math.cos(2)

-0.4161468365471424

2) Specifically importing `cos` from the math library. If we are using `cos` several times in the code, this can avoid repeated typing.

In [ ]:
from math import cos
cos(2)

-0.4161468365471424

3) Import everything in the module by using the \* wildcard. This can be an attractive option but is generally discouraged because different modules may use the same name for different functions. This method of importing imports not only the `cos` function, but all functions from the math library, for example `sqrt` and `tan`.

In [ ]:
from math import *
cos(2)

-0.4161468365471424

In [ ]:
sqrt(2)

1.4142135623730951

# Expressions

Before looking at other modules, let's look quickly at expressions. We have already seen several expressions in isolation (e.g. 18+5.5).

It is quite common the perform several operations as part of a single expression.

In [ ]:
a = 18 + 5.5 + 1
a

24.5

In this case behind the scenes Python adds 18 and 5.5 to get 23.5, it then adds 1 to get 24.5. In this case the order of the two operations does not matter, however in more complicated expressions the order can be important.

In [ ]:
a = 18*9**2/4
a

364.5

In [ ]:
a = 9/4**2*18
a

10.125

## Precedence

When there are two or more operations as part of an expression, we must figure out some way to determine which operation is performed first. We say that an operation that is performed fist is given *precedence* over the others.

Mathematical expressions in Python follow standard algebraic conventions:
1. Brackets
2. Exponents
3. Division/Multiplication
4. Addition/Subtraction

For example in the expression `1 + 2 * 3` the multiplication is done first, followed by the addition. 

In Python, as in algebra we can use brackets to prioritize an operation.

In [ ]:
1+2*3

7

In [ ]:
(1+2)*3

9

Most operations with equal precedence are evaluated left to right, again to mimic standard algebraic rules.  

One exception is exponents which are evaluated right to left, which again is how we typically think of exponents.

$4^{3^2} = 4^9$

In [ ]:
4**3**2

262144

In [ ]:
4**(3**2)

262144

In [ ]:
(4**3)**2

4096

Even though precedence rules are based on algebraic rules, they are enforced for any data type. 

In [ ]:
"a"*3+"b"

'aaab'

## Excersise

Write an expression that evaluates $3(7^2 + 4^{3^3} - 10)$.

Of course we can have much more complicated expressions involving the math module for example.

## Example

When boiling an egg, it has been determined that the time it takes for the center of the yolk to reach a desired temperature $T$ is given by

$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

where
* $M$ is the mass of the egg
* $\rho$ is the density
* $c$ is the specific heat capacity
* $K$ is the thermal conductivity
* $T_0$ is the temperature at $t=0$

$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

In [ ]:
T = 70 # desired temperature
M = 47
rho = 1.038
c = 3.7
K = 5.4e-3
T0 = 4

# compute time according to above formula
# we need the natural logarithm function and the pi constant from the math module
from math import log, pi

t = (M**(2/3)*c*rho**(1/3))/(K*pi**2*(4*pi/3)**(2/3))*log(0.76*(T0 - 100)/(T - 100))
print(t)

313.09454902221637


## Exercise 

A quadratic equation can be written as:
$$ ax^2 + bx + c = 0.$$

In general this equation has two (possibly equal) solutions and they are given by the formulas:

\begin{align*}
    x_1 = \frac{-b + \sqrt{b^2 - 4ac}}{2a},\\
    x_2 = \frac{-b - \sqrt{b^2 - 4ac}}{2a}.\\
\end{align*}

Use these formulas to compute the solutions of the equation $8x^2 + 16x + 4 = 0$.


In [ ]:
a = ...
b = ...
c = ...

x1 = ...
x2 = ...

## Calling Functions from Within Expressions

Function calls have high precedence. When multiple function calls are used in the same expression they are typically evaluated from left to right.

In [ ]:
person = "George Washington"
person.split()[1]

'Washington'

More complicated expressions are evaluated by first resolving commands inside parentheses.

In [ ]:
groceries = ["cereal", "milk", "apple"]
groceries.insert(groceries.index("milk") + 1, "eggs")
groceries

['cereal', 'milk', 'eggs', 'apple']

Here we first must evaluate `groceries.index("milk")` and then add 1 to it to find the index where we wish to insert "eggs".

## Exercise

Write a one line expression to capitalize the word to the immediate right of "milk" in the given list. Assume that you don't know the index of "milk" beforehand.

In [ ]:
groceries = ["cereal", "milk", "apple"]

# change apple to Apple
groceries[groceries.index("milk") + 1] = ...

# print new list
print(groceries)

['cereal', 'milk', Ellipsis]
